### PR TITLE
Issue #14233 - allow users to set a defualt value for ndivs for minor ticks on x and y axes

### LINKED ISSUE #14233: Feature Request: Allow setting default AutoMinorLocator
Via `rcParams` you can turn on minor ticks/grid for all your plots but there is currently no way to give a default interval for minor ticks to be shown.

AutoMinorLocator [defaults to 4 or 5](https://matplotlib.org/_modules/matplotlib/ticker.html#AutoMinorLocator) so potentially it could check to see if this value is set in rcParams and use that first instead?

### PR DESCRIPTION
**Pull request by KhadraMa**
## PR Summary
Pull request to resolve #14233 - Allow setting default AutoMinorLocator
I've added two new parameters to rcParams (rcsetup.py) and added a new case statement in the __call__ method of the AutoMinorLocator class (ticker.py). The functionality is as follows:

Allowed value of rcParams: {'auto', int}
Allowed value of Autominorticker(n): {'auto', int, None}
with:
- None: use rcParams
- int: use explicitly that many ticks
- 'auto': automatically choose a good number of ticks (the current 4/5 mechanism)

(functionality description pulled from timhoffm's comment from May 30, 2019)

Hi, this is my first pull request, so I'm sure there need to be some changes/updates before it's ready to be merged and I'm happy to make them. I'm a student at UTSC and while this is technically part of a course assignment I want to make clear that I'm personally interested in making a contribution and won't ghost you guys when the course ends (I know there's been some controversy about students contributing lately, but no need to worry).

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by QuLogic** [comment]
Something went wrong with your rebase; I don't think you wrote 634 commits.

**Comment by KhadraMa** [comment]
Yeah, I'm not sure what. I thought I followed the instructions exactly but I must have missed a step.

**Comment by KhadraMa** [comment]
I'll be honest, I don't know why the tests are failing. Is there any guidance or advice on how to improve the pull request so that the tests will pass?

**Comment by timhoffm** [comment]
Investigate the reasons for the failed test runs  by checking the *Details* links below . This might help: https://matplotlib.org/devdocs/devel/coding_guide.html#automated-tests

You might also want to try running the tests locally https://matplotlib.org/devdocs/devel/testing.html.

**Comment by jklymak** [comment]
@KhadraMa are you going to have time to come back to this or should we close?

**Comment by dopplershift** [comment]
@jklymak Seeing as this solves an open issue I'm not sure why we'd close?

**Comment by jklymak** [comment]
Because if no one is working on it, it doesn't make sense to leave open?  The original issue is still open, and this work is linked to it.  This was a student project, and after the semester they often do not come back to their PRs.

**Comment by KhadraMa** [comment]
Hi @jklymak, sorry for the late reply. I understand your concern - especially since I haven't been actively working on this for a few weeks now. The summer's been more hectic (and honestly more exhausting) than I expected. I still have every intention of finishing this though, so if it isn't too much trouble to keep it open I'd appreciate it. Thanks

**Comment by dopplershift** [comment]
@jklymak I don't know what philosophy for closing PRs has been discussed on the dev calls, so maybe my ideas are in opposition to the project writ large. IMO you only close PRs for being:
* Replaced
* Rejected ideas
* No longer applicable
* So woefully out of date that it's better to start from scratch

Yes it's linked from the issue, but that means people only stumble upon it if they're trying to fix *that* issue. I'd prefer leaving it open because, even if abandoned, it represents work that could go into matplotlib with a little more effort. My $0.02.

**Comment by jklymak** [comment]
@dopplershift perhaps there are better forums for this discussion?

**Comment by dopplershift** [comment]
@jklymak Probably. I've added the topic to the agenda for Monday's dev call.

File: lib/matplotlib/rcsetup.py [hunk 1]
```diff
@@ -775,6 +775,21 @@ def validate_hatch(s):
 validate_dashlist = _listify_validator(validate_floatlist)
 
 
+def validate_minor_tick_ndivs(n):
```
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:778]: This should probably defer to `validate_int_or_None` and then check that it's non-negative. And also should be private.

File: lib/matplotlib/rcsetup.py [hunk 2]
```diff
@@ -775,6 +775,21 @@ def validate_hatch(s):
...
+    """
+    Validate the given ndiv parameter for minor ticks.
+    ndiv can be either the string 'auto' or a non-negative integer.
+    """
+    if isinstance(n, int):
+        if (0 <= n):
+            return n
+        else:
+            raise RuntimeError(f'Value must be >=0; got {n}')
```
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:787]: Should be `ValueError`.

File: lib/matplotlib/rcsetup.py [hunk 3]
```diff
@@ -1287,6 +1302,8 @@ def _convert_validator_spec(key, conv):
     "xtick.minor.bottom":  validate_bool,      # draw bottom minor xticks
     "xtick.major.top":     validate_bool,      # draw top major xticks
     "xtick.major.bottom":  validate_bool,      # draw bottom major xticks
+    "xtick.minor.ndivs": ['auto', validate_minor_tick_ndivs],
```
Comment (QuLogic) [inline lib/matplotlib/rcsetup.py:1305]: I don't think this actually works the way you want, does it?

File: lib/matplotlib/tests/test_ticker.py [hunk 4]
```diff
@@ -191,6 +191,70 @@ def test_additional(self, lim, ref):
 
         assert_almost_equal(ax.yaxis.get_ticklocs(minor=True), ref)
 
+    def test_ndivs_rcParams_auto(self):
+        # when no parameter is given rcParams should be used
+        # auto
+        mpl.rc_context(rc={'xtick.minor.ndivs': 'auto'})
+        mpl.rc_context(rc={'ytick.minor.ndivs': 'auto'})
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_ticker.py:198]: This needs to be used as a context manager, or it has no effect. Or you can set `rcParams` directly; they'll be reset at the end of the test.

File: lib/matplotlib/tests/test_ticker.py [hunk 5]
```diff
@@ -191,6 +191,70 @@ def test_additional(self, lim, ref):
...
+                               0.5, 0.55, 0.65, 0.7, 0.75, 0.85, 0.9,
+                               0.95, 1.05, 1.1, 1.15, 1.25, 1.3, 1.35])
+        assert_almost_equal(ax.xaxis.get_ticklocs(minor=True), test_value)
+        assert_almost_equal(ax.yaxis.get_ticklocs(minor=True), test_value)
+
+    def test_ndivs_rcParams_int(self):
+        # when no parameter is given rcParams should be used
+        # non-neg int
+        mpl.rc_context(rc={'xtick.minor.ndivs': 2})
+        mpl.rc_context(rc={'ytick.minor.ndivs': 2})
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_ticker.py:215]: Ditto.

File: lib/matplotlib/ticker.py [hunk 6]
```diff
@@ -2933,6 +2934,12 @@ def __call__(self):
             return []
 
         if self.ndivs is None:
+            if self.axis.__name__ == 'xaxis':
+                self.ndivs = mpl.rcParams['xtick.minor.ndivs']
+            else:
+                self.ndivs = mpl.rcParams['ytick.minor.ndivs']
```
Comment (QuLogic) [inline lib/matplotlib/ticker.py:2940]: This won't work for 3D; it normally uses the *x* setting for *z*.
Also, better to use `self.axis.axis_name` instead.

**Comment by IlievskiV** [comment]
@KhadraMa is this PR still relevant and opened to work on? If yes, may I take it over and try to complete it?

**Comment by KhadraMa** [comment]
@IlievskiV go for it. With the way things are going it would be a while before I have time to work on this again - but it's open and should be finished so yeah, have at it

**Comment by jklymak** [comment]
Thanks @KhadraMa I'll close this version....

**Pull request closed without merging**